Sequence Templates

JosephVolosin opened the pull request:
This PR implements the UI for sequence templates (see [`aerie` PR](https://github.com/NASA-AMMOS/aerie/pull/1638) for more information)

# New Environment Variable
This feature introduces a new environment variable called `PUBLIC_SEQUENCING_MODE` which is set to either `templating` or `expansion` depending on the feature the user wants. Command Expansion and Sequence Templating are two ways to get the same outputs, so the user must decide when they build their instance which they'd like to use.

# Sequence Template Editor
To create/edit/delete sequence templates, there is a new route `/sequence-templates` that brings the user to an editor based on the sequence editor in `/sequencing`. 

![image](https://github.com/user-attachments/assets/007f2b25-dcda-4af6-89a2-562d7607a853)

The editor window utilizes `CodeMirror` and implements an 'overlay' language to support [Handlebars](https://handlebarsjs.com/) formatting. Additionally, the same parcel/dictionary capabilities from the sequence editor are available for linting, formatting, auto-complete, etc.

# Plan View - Sequences & Expansion
Within the Plan view, there is a new tab called `Sequences & Expansion` which is used to create sequences and expand them using either command expansion _or_ sequence templating. The functionality from the previous command expansion panel has been rolled into this panel. New sequences and sequence filters can be created through the `New` button next to the table filter.

![image](https://github.com/user-attachments/assets/6148faa3-3b57-4164-b5fb-7091c11436d2)

The table shows both expansion sequences and sequence filters. Expansion sequences can be deleted, expanded (using command expansion or sequence templating), and have the expanded output viewed if it exists.

![image](https://github.com/user-attachments/assets/e767acaa-6d10-4cf2-a0a6-1656d362652e)

Sequence filters can be deleted, modified after creation, and applied. 'Applying' a sequence filter generates a new expansion sequence with spans that match the filter criteria automatically attached to the sequence ID.

![image](https://github.com/user-attachments/assets/1bd6f253-ce42-4332-89ed-54d296bd64dd)

# Step-by-step guide
0. Upload/Create a: Model, Plan, Dictionary, and Parcel
1. Navigate to `Sequence Templates` (`/sequence-templates`) through the Aerie logo in the top-left
2. Click `New Template` and enter a 1) Template Name, 2) Template Language (`STOL`, `SEQN`, or `TEXT`), 3) Parcel Id, 4) Model Id, 5) Activity Type. Click `Create Template`
3. Click the template in the table and then write expansion instructions in the center panel (ex: `/C FSW_CMD_0 "ON" false 1`). Click `Save` or use `Ctrl+S` to save the template.
4. Navigate to a plan through the Aerie logo in the top-left & selecting a plan from the table
5. Add at least one activity directive to the timeline and run a simulation
6. Change one of the UI panels to the `Sequences & Expansion` panel
7. Click `New` and create a new sequence filter. In the filter builder modal, enter a name and at least one filter constraint that will capture the previously added activity directive
8. Hover over the sequence filter in the table and click the `Create Sequence` button that appears, select a time range in the modal that appears that includes the previously added activity directive (or, leave it as the default - the plan time range)
9. Hover over the expansion sequence that is generated and click `Expand Sequence`
10. Hover over the same expansion sequence and click `Show Expanded Sequence`

JosephVolosin committed:
Add activity filtering to sequence creation

JosephVolosin committed:
Move sequence creation to a modal

Another approach to do the same as the last commit - some issues w. displaying the ActivityFilterBuilder from within the modal

JosephVolosin committed:
Create Sequencing panel in plan view

JosephVolosin committed:
Draft of rigging templating button

JosephVolosin committed:
Add bulk functionality to SequencingPanel

JosephVolosin committed:
Refactor SequenceDefinition => SequenceFilter; draft work for templating actions

Commit:
Debugging naive Handlebars overlay

Debugging a naive overlay approach to integrating handlebars into the editor.

Commit:
Fix autocompletion

Fix Seq-language autocompletion to work outside, and only outside, of template insert fields.

Commit:
Ignore symlink to aerie

Commit:
WIP - building sequence template editor

JosephVolosin committed:
WIP UI work for expanded sequences

JosephVolosin committed:
WIP UI working for sequencing in plan, expanded templates, etc.

Commit:
mark expansions as dirty on timeline change; remove ephemeral sequence generation

Commit:
WIP - building sequence template editor

Commit:
WIP - adding "new template" functionality

Doesn't include any back-end support, but this does stub out the back end as a store that should be easy-ish to replace with a real version.

Commit:
WIP - add "new template" modal

Commit:
WIP - use editor change event to save template content

Commit:
WIP - debug parcel selection

Commit:
WIP - stop cursor jumping in editor

JosephVolosin committed:
WIP re-organization of sequencing panel, back-end linkage work

JosephVolosin committed:
WIP debugging template expansion

Commit:
fix with rebase

Commit:
fix imports

Commit:
WIP - figuring out handlebars autocomplete

Commit:
WIP - update handlebars completions

Commit:
WIP - add activity + model to template draft metadata

JosephVolosin committed:
Improve sequence template creation modal

Adds dropdowns and selections for parcle, model, and activity type selections

JosephVolosin committed:
Connect sequence template front-end to back-end

JosephVolosin committed:
Selected sequence details

JosephVolosin committed:
/sequencing/templates => /sequence-templates

JosephVolosin committed:
WIP - Re-design of tabs to sequences & sequence template panels

JosephVolosin committed:
Linting/formatting update

JosephVolosin committed:
Cleanup cont.

JosephVolosin committed:
Move sequence template types/stores => sequence-template.ts

JosephVolosin committed:
Continued cleanup & moving to type/store for sequence-template

mattdailis committed:
Use graphql action instead of insert to create template

JosephVolosin committed:
Unittesting new effects functions

JosephVolosin committed:
Unittesting pt.2

JosephVolosin committed:
Refactor of Sequence Template page & editor

Rips out some sequence adaptation related code that is not related to the current iteration of sequence templates.

JosephVolosin committed:
Add missing , for linting

Commit:
implement feature switch between templating and legacy expansion

Commit:
roll back changes related to dirty bit handling

Commit:
update expansion button in plan menu

Commit:
merge all expansion into one panel with tabs, using feature switch

JosephVolosin committed:
WIP UI re-design, creating SequencesAndExpansionPanel

JosephVolosin committed:
Cleanup

Remove old comments/console.logs, move 'feature unavailable' errors to written stores, add missing GQL permissions

duranb commented on .gitignore:
```diff
@@ -1,4 +1,5 @@
 .DS_Store
+aerie
```
Is there a reason to add this? There shouldn't be an `aerie` directory in the UI

JosephVolosin replied:
Reverted to the `develop` version - think this got leftover from testing!

duranb commented on package.json:
```diff
@@ -54,6 +54,7 @@
     "@sveltejs/adapter-node": "5.0.1",
     "@sveltejs/kit": "^2.5.4",
     "@tanstack/svelte-virtual": "^3.11.2",
+    "@xiechao/codemirror-lang-handlebars": "^1.0.4",
```
Was this the best library for the feature? I'm not seeing much documentation on it and I'm not finding their github repo. Ideally, we'd like to only utilize libraries that can be peer reviewed and vetted.

JosephVolosin replied:
That's a fair point, I do think this was the only repo we could find that had support for Handlebars in CodeMirror - @mattdailis any other thoughts?

mattdailis replied:
That is fair. It's the best we were able to find, so I think the only real alternative would be to fork it and own the code ourselves. It looks pretty small, so I don't think that's out of the question. Would that address your concern, Bryan, or am I missing the point?

duranb replied:
> That is fair. It's the best we were able to find, so I think the only real alternative would be to fork it and own the code ourselves. It looks pretty small, so I don't think that's out of the question. Would that address your concern, Bryan, or am I missing the point?

A fork would be fine! I just would like us to be confident that the libraries that we use don't have hidden tracking or other nefarious tactics.

JosephVolosin replied:
We ended up rolling the language definition directly into `aerie-ui`, see [this commit](https://github.com/NASA-AMMOS/aerie-ui/pull/1642/commits/a044e6d2509791f44f738f165289b99f84586208)

mattdailis committed:
Remove parcelId from ExpandTemplates mutation

mattdailis commented on src/stores/sequencing.ts:
```diff
@@ -12,13 +14,27 @@ import {
 
+/* Environment Configuration */
+export const sequenceExpansionMode: Writable<SequencingMode> = writable(
+  env.PUBLIC_SEQUENCING_MODE === 'templating' ? SequencingMode.EXPANSION : SequencingMode.EXPANSION,
```
```suggestion
  env.PUBLIC_SEQUENCING_MODE === 'templating' ? SequencingMode.TEMPLATING : SequencingMode.EXPANSION,
```

mattdailis commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,384 @@
+    </ActivityFilterBuilder>
+    <div class="sne-controls">
+      <div class="sne-filter">
+        <input
```
This input gets out of sync with the input in the filter builder modal. E.g.:
1. Fill in the text box next to the new filter button
2. Make a filter, but don't give it a name
3. Save the filter

The created filter doesn't have the name from step 1, and the text box is still populated.

I think what I expect is that the text box should clear after I click "new". If I'm creating a sequence filter, that value should be used to seed the name of the filter in the text box in the modal.

mattdailis replied:
I may have misunderstood the intent of that text box - it says "Filter text", but I guess it means "filter the items in this list so I can find what I'm looking for" rather than "this is the name of the SequenceFilter"

JosephVolosin replied:
Ha, ah yeah that is confusing now that I think about it.. that box (which is unfortunately next to that 'New' button) is meant to filter the results in the table underneath. The name of the sequence filter you're creating or viewing is **just** bound to that entry in the filter builder. Maybe we should update the text in the filter box to say 'Filter items...'?

JosephVolosin committed:
Revert .gitignore

Aerie sub-dir should not nominally exist

JosephVolosin committed:
Phase out parcel selection, general cleanup

duranb commented on src/types/sequencing.ts:
```diff
@@ -22,6 +22,18 @@ import type { SequenceTypes } from '../enums/sequencing';
 import type { GlobalType } from './global-type';
+import type { ActivityLayerFilter } from './timeline';
+
+export type SequenceActivityFilter = ActivityLayerFilter;
```
Would it be correct to say that this is the same as an `ActivityLayerFilter`? The purpose of the `ActivityLayerFilter` is more geared towards activities.

JosephVolosin replied:
Yep - I made `SequenceActivityFilter` assuming that it would be somewhat different from `ActivityLayerFilter`, but that didn't really end up being the case so I've re-factored everything back to using `ActivityLayerFilter`.

duranb replied:
Are you using the SequenceActivityFilters in the timeline somewhere? It looks like you're utilizing the "filter" part of the `ActivityLayerFilter`. If it's not really a timeline feature, then I think it would be less confusing to maybe rename and rework `ActivityLayerFilter` and its surrounding types to something more reusable for other features by using generic types.

JosephVolosin replied:
The filtering for this PR doesn't interact with the timeline itself at all, so I think we could definitely change `ActivityLayerFilter` to a more generic name for both the expansion use and timeline use - the filter only revolves around activity types/filters so maybe just `ActivityFilter`?

duranb replied:
Yeah! That might work. My concern is that because it isn't timeline related, this would cause unnecessary refactoring for sequence templates should the format of the timeline filters need to change. I think it would only be appropriate to reuse `ActivityLayerFilter` or portions of it if there was some common functionality or it was generic enough where a refactor in one feature (sequence templates or the timeline) should naturally result an update in the other (if that makes sense).

JosephVolosin replied:
Ah, that does make sense - so instead would our best choice here be to keep the types separate (`ActivityLayerFilter` and something like `SequenceActivityFilter`)? And both would be defined completely independent from each other

duranb replied:
Yeah something like that might be more ideal. There might be more atomic types that are less likely to change like `FilterOperator` or `DynamicFilterDataType` that could be pulled out into a generic `filter.ts` type file and reused just so that everything isn't duplicated.

JosephVolosin replied:
Ended up reorganizing a few types including `FilterOperator`, `ActivityLayerDynamicFilter`, `DynamicFilterDataType`, and `ActivityLayerFilter` - let me know what you think, I can roll-back/change if you think this is too much/too little, etc.

duranb commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,373 @@
+      <div class="sne-buttons">
+        <button
+          class="st-button secondary new-button"
+          style="position: relative; z-index: 1"
```
Do you mind avoiding static inline styles? It makes it difficult to maintain when the styles are spread out throughout the component.

JosephVolosin replied:
Removed this since I don't believe it's needed!

duranb commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,373 @@
+
+  $: sequencesAndFilters = [...$expansionSequences, ...$sequenceFilters];
+
+  // TODO: Better cleanup of activeSequenceFilterName and activeSequenceFilterId
```
Do you mind elaborating more as to what needs to be cleaned up for `activeSequenceFilterName` and `activeSequenceFilterId`?

JosephVolosin replied:
Removed this TODO now - these weren't getting cleaned up properly between uses of the 'FilterBuilder', but the addition of a reset underneath `on:visibilityChange` fixes this!

duranb commented on src/components/modals/ExpansionSequenceModal.svelte:
```diff
@@ -18,11 +21,21 @@
+      expandedTemplate => expandedTemplate.seq_id === expansionSequence.seq_id,
+    );
+    outputStr = expandedTemplate?.expanded_template ?? `No output found for sequence "${expansionSequence.seq_id}"'`;
+    // TODO: Language - see question in Notes
```
Do you mind pasting the question directly here if you intend to keep this TODO?

JosephVolosin replied:
Removed - was a reminder to myself that I forgot to cleanup!

AaronPlave commented on src/components/timeline/form/TimelineEditor/ActivityFilterBuilder.svelte:
```diff
@@ -203,20 +212,14 @@
   $: if (appliedFilter) {
-    const seenSpans: Record<number, boolean> = {};
-    let count = appliedFilter.directives.length;
+    filteredSpans = [];
```
Reason for this change?

JosephVolosin replied:
This was originally for when we thought we were going to pull the spans through the UI itself, though we now moved it out to the back-end - I went ahead and reverted this section back to what it is on `develop`!

JosephVolosin committed:
Cleanup and e2e testing start

Sequence Template e2e tests working

JosephVolosin committed:
First iteration of PR comment cleanup

JosephVolosin committed:
Fix broken imports in SequenceTemplateEditor

Resolves issues from rebase/re-org of codemirror + sequencing related files

JosephVolosin committed:
Prettier cleanup

JosephVolosin committed:
Fix leftover placeholder aria-labels

JosephVolosin committed:
WIP E2E testing

mattdailis committed:
Add toggle

JosephVolosin committed:
E2E testing

JosephVolosin commented on src/utilities/effects.ts:
```diff
@@ -3691,25 +3924,19 @@ const effects = {
   ): Promise<string | null> {
     try {
-      const data = await reqHasura<GetSeqJsonResponse>(
+      const data = await reqHasura<any>(
```
Fix `any` here - I think `GetSeqJsonResponse` might have changed slightly which caused this to change to `any` temporarily.

JosephVolosin replied:
Fixed! Using `ExpandedSequence[]` now.

JosephVolosin committed:
Cleanup, fix stale selection on sequence templates

mattdailis committed:
Set SEQUENCING_LANGUAGE in tests

mattdailis committed:
Split seq-n and seq-n-handlebars

JosephVolosin committed:
Add Handlebars CodeMirror language support

pranav-super commented on .env:
```diff
@@ -7,5 +7,6 @@ PUBLIC_HASURA_SERVER_URL=http://localhost:8080/v1/graphql
 PUBLIC_HASURA_WEB_SOCKET_URL=ws://localhost:8080/v1/graphql
 PUBLIC_AUTH_SSO_ENABLED=false
 PUBLIC_TIME_PLUGIN_ENABLED=false
+PUBLIC_SEQUENCING_MODE=templating
```
Whether this is the default mode is something being deliberated on AERIE main (see [here](https://github.com/NASA-AMMOS/aerie/pull/1638#discussion_r1999083732)).

JosephVolosin replied:
Looks like we settled this over there and Matt updated it here!

pranav-super commented on src/components/menus/AppMenu.svelte:
```diff
@@ -100,7 +113,20 @@
-    <MenuItem on:click={() => showAboutModal()}>
+    <MenuItem
+      on:click={e => {
+        if (e.detail.shiftKey) {
```
Should we keep this switch on? I think having this functionality available somewhere in the UI is helpful, but it seemed like from an earlier conversation, the decision was to delegate this to a configuration setting and to leave it at that. Is this something we mean to leave in?

JosephVolosin replied:
We had a long conversation about this and decided the best way to do it, for now, is through a hidden switch in the UI like this. Hopefully though in the future we'll be able to have a better way to allow the E2E tests to get to both Typescript & Templating sides of expansion

pranav-super commented on src/components/modals/ExpansionSequenceModal.svelte:
```diff
@@ -18,11 +21,20 @@
+    outputStr = expandedTemplate?.expanded_template ?? `No output found for sequence "${expansionSequence.seq_id}"'`;
+  } else {
+    effects
+      .getExpansionSequenceSeqJson(expansionSequence.seq_id, expansionSequence.simulation_dataset_id, user)
```
Question to @mattdailis and @JosephVolosin:

This makes use of the `expanded_sequences` table, which makes sense - have usages of the `getSequenceSeqJson` endpoint been fully phased out in the UI at this point? It seems like we can do away with that endpoint entirely, especially if the UI has demonstrated that it can be replaced with a call to the `expanded_sequences`.

JosephVolosin replied:
What is the reason we're moving away from `getSequenceSeqJson`? FWIW, pulling directly from `expanded_sequences` (as defined currently) has the unexpected result that if you've re-run expansion within the same simulation dataset, you end up with an array of results. We _could_ filter that down to only provide the latest one via. the `creation_time` field, but just want to make sure we really do want to swap this functionality

JosephVolosin replied:
Talked w. Pranav and Matt on this - querying `expanded_sequences` is the **correct** way to go forward for this PR, and we are OK grabbing the latest expansion based on creation time. Updated on the UI!

pranav-super commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,366 @@
+                  <FilterIcon />
+                </button>
+              </div>
+              <div use:tooltip={{ content: 'Apply Filter', placement: 'top' }}>
```
This button doesn't do anything if the plan is empty. So creating a filter and then a sequence from that filter with an empty plan, before adding activities, is not possible.

That shouldn't be an issue but might be worth noting, especially if filter reapplication and such become desired behavior later?

JosephVolosin replied:
Fair question - in general, I'm not 100% certain how we want to handle situations where the action will 'fail' because there is no initial simulation

pranav-super replied:
Pushed a fix just now that comprehensively disables buttons in these cases.

pranav-super commented on src/components/modals/TimeRangeModal.svelte:
```diff
@@ -0,0 +1,71 @@
+</script>
+
+<Modal {height} {width}>
+  <ModalHeader on:close>Apply Filter To Range</ModalHeader>
```
Maybe this should have a more intuitive title like "Create Sequence from Filter", and then "Apply Filter to Range" could be the body text? Just to make it clear that in applying a filter we are creating a sequence?

JosephVolosin replied:
Agreed, I swapped the header content around and added some text under:
![image](https://github.com/user-attachments/assets/5c7b1fe9-806a-47f5-947f-2eb64557ab2f)

pranav-super reviewed (commented):
A few minor comments and questions. Everything looks fine otherwise!

Commit:
change default sequencing mode to legacy

Commit:
addTemplateExpansion -> addTemplate

cartermak commented on src/enums/sequencing.ts:
```diff
@@ -2,3 +2,8 @@ export enum SequenceTypes {
 }
+
+export enum SequencingMode {
+  LEGACY,
```
```suggestion
  EXPANSION,
```

cartermak commented on src/stores/sequencing.ts:
```diff
@@ -12,13 +14,27 @@ import {
 
+/* Environment Configuration */
+export const sequenceExpansionMode: Writable<SequencingMode> = writable(
+  env.PUBLIC_SEQUENCING_MODE === 'legacy' ? SequencingMode.LEGACY : SequencingMode.TEMPLATING,
```
This sets `SequencingMode.LEGACY`, but all the UI elements check for `SequencingMode.EXPANSION`.

This fix go the other way, too -- but there are lots of uses of `SequencingMode.EXPANSION` and only one usage of `SequencingMode.LEGACY`. Not sure if there's one term we prefer.

```suggestion
  env.PUBLIC_SEQUENCING_MODE === 'legacy' ? SequencingMode.EXPANSION : SequencingMode.TEMPLATING,
```

cartermak commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,366 @@
+                  <JournalCodeIcon />
+                </button>
+              </div>
+              <div use:tooltip={{ content: 'Expand Sequence', placement: 'top' }}>
+                <button
+                  aria-label={`Expand '${sequenceOrFilter.seq_id}'`}
+                  class="st-button icon"
+                  on:click|stopPropagation={() => {
+                    if (isExpansionSequence(sequenceOrFilter)) {
+                      onExpandSequence(sequenceOrFilter);
+                    }
+                  }}
+                >
+                  <PlayIcon />
+                </button>
+              </div>
```
It's misleading that each sequence gets an "Expand Sequence" button: it seems like it'll only expand that single sequence. Is there a reason why we moved away from a single button to expand the entire plan?

JosephVolosin replied:
I'll be honest I think I may have biased this UI decision more towards GSFC use because we generally don't have the idea of 'multiple sequences' - instead one sequence representing our whole load makes more sense. However, I think a 'Run All' button definitely makes sense _especially_ if you're using Typescript/Legacy/Command Expansion mode - so I'm okay with adding it back in.

cartermak replied:
OK nice, yeah I totally see how we ended up there. I think, though, that even in the case of template expansion, we expand all sequences at once? I think with either legacy or template expansion, one "run" will affect all sequences?

JosephVolosin replied:
Added an `Expand All` button for _both_ modes - the row-specific 'expand' buttons will still _only_ expand that sequence, but `Expand All` will do all that are configured for the current/latest simulation dataset

cartermak replied:
I'm happy with it as is, but for what it's worth - in legacy CE mode, the row-specific 'Expand' button still expands all sequences: https://github.com/NASA-AMMOS/aerie-ui/blob/cf048d17a5ac3045afb0c695c6ef0f7e767560fa/src/components/expansion/SequencesAndExpansionPanel.svelte#L142

cartermak reviewed (commented):
A couple comments from a first pass today. I spent a while trying to debug a crash when I try to view an expanded sequence in legacy expansion mode but cut myself off before spending my entire evening on it.

JosephVolosin commented:
@AaronPlave / @duranb, with our new panel for `Sequences & Expansion` (and retiring the `Expansion` panel), what (if anything) needs to change with the view JSON schema?

mattdailis committed:
Rename PUBLIC_SEQUENCING_MODE to PUBLIC_COMMAND_EXPANSION_MODE and use the word typescript instead of legacy

JosephVolosin committed:
Move 'LEGACY' to 'TYPESCRIPT' for sequence mode flag; E2E test fixes

cartermak commented:
Edit: since been resolved!

Some things I've observed but don't yet understand:

1. I can sometimes get the UI into a state where the app menu shows me neither expansion nor templates as an option:
![image](https://github.com/user-attachments/assets/afe046e2-cc85-4467-8b85-e19b8c800e09)

2. For some reason, the `input` value passed to `seqJsonToSequence` when viewing an expansion run has changed from a stringified JSON to a parsed object. I can't figure out why, because it doesn't look like the underlying stores or processing changed at all. Unless this is broken on develop, too?

JosephVolosin committed:
Improve text on TimeRangeModal

JosephVolosin commented:
> Some things I've observed but don't yet understand:
> 
>     1. I can sometimes get the UI into a state where the app menu shows me neither expansion nor templates as an option:
>        ![image](https://private-user-images.githubusercontent.com/49076171/425226428-afe046e2-cc85-4467-8b85-e19b8c800e09.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1Njc5MzksIm5iZiI6MTc0MjU2NzYzOSwicGF0aCI6Ii80OTA3NjE3MS80MjUyMjY0MjgtYWZlMDQ2ZTItY2M4NS00NDY3LThiODUtZTE5YjhjODAwZTA5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIxVDE0MzM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljOWZlZmNlMmNiOTNmMTU0M2JjMDU1YTY2YmEyZTVjOWZmZmY1ZTlmMTIyZmRhZWUzN2YwNWMzMzFmMmRjMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jij_0Bv-JPlHB9C5PReE8gSCoDn0p3uQ8uF8T4Jz_Js)
> 
>     2. For some reason, the `input` value passed to `seqJsonToSequence` when viewing an expansion run has changed from a stringified JSON to a parsed object. I can't figure out why, because it doesn't look like the underlying stores or processing changed at all. Unless this is broken on develop, too?

1. Can you try to replicate? I _think_ this was being caused by missing a few updates in the UI after we switched around some naming related to the env variable.
2. I think this is because we moved from calling a `getSequenceSeqJson` to directly referencing `expanded_sequences`. See convo here: https://github.com/NASA-AMMOS/aerie-ui/pull/1642#discussion_r2001832572

JosephVolosin committed:
Add 'Expand All' button, fix GQL query for retrieving expanded sequences

JosephVolosin committed:
Disable expansion buttons if using 'Typescript' expansion and no expansion set is selected

JosephVolosin committed:
Lint/format fixes

JosephVolosin committed:
Fix 'any' => 'ExpandedSequence[]' in getExpansionSequenceSeqJson

JosephVolosin committed:
E2E test fixes

duranb commented on e2e-tests/fixtures/ExpansionRules.ts:
```diff
@@ -111,6 +114,11 @@ export class ExpansionRules {
     await this.page.goto('/expansion/rules', { waitUntil: 'networkidle' });
     await this.page.waitForTimeout(250);
     await expect(this.rulesNavButton).toHaveClass(/selected/);
+    const typescriptIsToggledOff = await this.page.getByText('COMMAND_EXPANSION_MODE').isVisible();
```
What specifically is this trying to select? Is this some button or label? It's not immediately clear where this value is being placed in the HTML and I worry that it'll make it hard to debug later on if the text ever changes.

JosephVolosin replied:
When the user (or in this case the test) accesses a page that isn't available for their current sequencing mode (ex: accessing `/sequence-templates` when you're using the `typescript` sequencing mode) it has text on the page (stored within `$commandExpansionNotAvailable` or `$templatingNotAvailable`) telling the user/test the page isn't available. `COMMAND_EXPANSION_MODE` is directly referenced in the text since it's the environment variable used to define the mode, but I could expand this `getByText` to look for more of the error. I was also trying to use a piece that would appear in both messages to simplify

duranb replied:
Got it. Yeah I think including more of the error here would help and maybe renaming the variable to be something like `typescriptNotEnabledError` might make it easier to understand the purpose for it. Also, should it say `PUBLIC_COMMAND_EXPANSION_MODE` to match the env var?

cartermak replied:
Resolving since it looks like this check was removed altogether - feel free to re-open if you feel otherwise

duranb commented on src/utilities/sequence-editor/languages/handlebars/handlebars.ts:
```diff
@@ -0,0 +1,17 @@
+import { foldInside, foldNodeProp, indentNodeProp, LRLanguage } from '@codemirror/language';
```
Is handlebars itself a standalone language and is seq-n-handlebars a sub-language?

JosephVolosin replied:
Handlebars is a standalone language and seq-n-handlebars is the grammars of both the Handlebars language and the SeqN language, defined with the Handlebars grammar as an [overlay](https://lezer.codemirror.net/docs/ref/) (see 'Mixed Parsing' section)

duranb replied:
Oh nice! Can this be moved to a sub directory of `handlebars`?

cartermak replied:
@duranb are you happy with this, or did you still want this file moved? I'm not sure I understand the suggestion

duranb replied:
Oh sorry. My suggestion was because `seq-n-handlebars` is dependent on `handlebars`, that it should be moved under `handlebars` if it isn't a standalone language.

JosephVolosin replied:
My stance would be to keep `seq-n-handlebars` as a separate directory since it _is dependent_ on `handlebars` but its also dependent on `seq-n`

mattdailis replied:
I chatted about this with Bryan yesterday - let's leave them as separate directories for now. If we find that we need to make it more obvious that handlebars is meant to be used differently from, say, SeqN, we could move it under a directory `templating-languages/handlebars`, but let's leave it where it is for the time being.

Commit:
update disabled button behavior in SequencesAndExpansionPanel.svelte

duranb commented on docs/ENVIRONMENT.md:
```diff
@@ -7,6 +7,7 @@ This document provides detailed information about environment variables for Aeri
 | `ORIGIN`                         | Url of where the UI is served from. See the [Svelte Kit Adapter Node docs][svelte-kit-adapter-node-docs].     | `string` | http://localhost                 |
 | `PUBLIC_AERIE_FILE_STORE_PREFIX` | Prefix to prepend to files uploaded through simulation configuration.                                         | `string` | /usr/src/app/merlin_file_store/  |
 | `PUBLIC_AUTH_SSO_ENABLED`        | Whether to use the SSO-based auth flow, or the /login page auth flow                                          | `string` | false                            |
+| `PUBLIC_COMMAND_EXPANSION_MODE`  | `typescript` or `templating`                                                                                  | `string` | typescript                       |
```
Can this have more of a description?

cartermak commented on e2e-tests/tests/plan-sequences-and-expansion.test.ts:
```diff
@@ -0,0 +1,199 @@
+  // await plans.deletePlan();
+  // await models.goto();
+  // await models.deleteModel();
+  // await parcels.goto();
```
Either delete plan & model or clean these up?

JosephVolosin replied:
Think I forgot a commit - this was leftover from testing but is now fixed as of the latest e2e test fixes!

cartermak commented on e2e-tests/tests/plan-sequences-and-expansion.test.ts:
```diff
@@ -0,0 +1,199 @@
+    await plan.sequenceExpansionOutputModal.waitFor({ state: 'attached' });
+    await plan.sequenceExpansionOutputModal.waitFor({ state: 'visible' });
+    await page.getByText('Loading Editor...').waitFor({ state: 'detached' });
+    // await expect(plan.sequenceExpansionOutputModal.getByText(sequenceTemplateContent)).toBeVisible();
```
Is there a reason this final check is commented out?

JosephVolosin replied:
Same as https://github.com/NASA-AMMOS/aerie-ui/pull/1642#discussion_r2015017961.. also cleaned up now though and checking a real condition!

duranb commented on src/stores/sequencing.ts:
```diff
@@ -12,13 +14,27 @@ import {
+);
+
+export const commandExpansionNotAvailable: Writable<string> = writable(
+  'Command Expansion functionality not available in "templating" sequencing mode. Please change the COMMAND_EXPANSION_MODE environment variable to "typescript" in docker-compose.yml.',
```
Should "COMMAND_EXPANSION_MODE" be "PUBLIC_COMMAND_EXPANSION_MODE"?

duranb replied:
These stores don't seem to be being written to other than on initialization. Do these need to be stores or can they just be constants?

JosephVolosin replied:
Good point - created `constants/command-expansion.ts` to hold these 3 stores as constants instead

duranb replied:
Thanks! One last nitpick about this. Can you rename the constant variable names to have "_MESSAGE" or "_ERROR" at the end just so it's more explicit as to what it is for?

JosephVolosin replied:
Yep! Added "_MESSAGE" to both of them

duranb commented on src/stores/sequencing.ts:
```diff
@@ -27,8 +43,11 @@ export const parsedCommandDictionaries: Writable<Record<string, AmpcsCommandDict
 export const parsedParameterDictionaries: Writable<Record<string, AmpcsParameterDictionary>> = writable({});
 
-/* Subscriptions. */
+export const selectedParcel: Writable<number | null> = writable(null);
```
Can this be renamed to `selectedParcelId`?

JosephVolosin replied:
Removing this store completely - this is actually leftover from before we realized we don't need to select a parcel at expansion-time. Instead they're only used within the editor, and the parcel is tied directly to the template when you create it

duranb commented on src/stores/sequence-template.ts:
```diff
@@ -0,0 +1,53 @@
+
+/* Subscriptions. */
+
+export const activityTypes = gqlSubscribable<ActivityType[]>(gql.SUB_ACTIVITY_TYPES, { modelId }, [], null);
```
This store already exists in the `plan.ts` store file. Can we pull this out into its own file so that it isn't duplicated?

JosephVolosin replied:
I copied it out here since the `modelId` is different between `sequence-template.ts` and `plan.ts` (set via. the dropdown vs. set by the current plan). Whats the best way to let it be accessed via. either use of `modelId`?

duranb replied:
Yeah that makes sense. I think then both the plan and sequence-template stores need to at least be renamed for their respective features (e.g. `planActivityTypes`, `planModelId`) to help disambiguate.

JosephVolosin replied:
Refactored both `modelId` and `activityTypes` in both stores!

cartermak reviewed (commented):
My one outstanding prior comment turned out to be an existing bug, so capturing separately as #1655 since it's not the fault of this PR!

duranb commented on src/components/expansion/SequencesAndExpansionPanel.svelte:
```diff
@@ -0,0 +1,404 @@
+    hasDeletePermissionSequenceFilter = featurePermissions.sequenceFilter.canDelete(user, $plan.model);
+  }
+
+  $: relevantSimulationDatasetIds = $simulationDatasetsPlan?.map(dataset => dataset.dataset_id);
```
Do you mind declaring the types for these reactive variables before they're set?

JosephVolosin replied:
Added!

JosephVolosin committed:
Use correct ID for finding plan-related simulation dataset IDs

JosephVolosin committed:
Separate typescript & templating tests

duranb committed:
Sequence template e2e test running (#1656)

add a separate e2e test suite for sequence templates

JosephVolosin committed:
Linting

JosephVolosin committed:
Remove unused selectedParcel store

JosephVolosin committed:
Move sequencing mode stores to constants

duranb committed:
Run all the test suites as separate projects in single playwright config (#1658)

run all the test suites as separate projects on different server ports in single playwright config

JosephVolosin committed:
Allow sequence template editor to use full height

JosephVolosin committed:
Declare reactive variables

JosephVolosin committed:
Refactor modelId and activityTypes stores from sequence-templates + plan

JosephVolosin committed:
Set proper 'modelId' arg in store

JosephVolosin committed:
Remove hidden toggle-related E2E tests

JosephVolosin committed:
Rename modelId => planModelId in plan.ts Mock

JosephVolosin committed:
Update description for PUBLIC_COMMAND_EXPANSION_MODE

duranb commented on src/stores/plan.ts:
```diff
@@ -52,13 +52,18 @@ export const plan: Readable<Plan | null> = derived([initialPlan, planMetadata],
+);
 
-export const subsystemTags: Readable<Tag[]> = derived(activityTypes, $activityTypes => {
+export const subsystemTags: Readable<Tag[]> = derived(planModelActivityTypes, $activityTypes => {
```
Do you mind also updating `$activityTypes` to `$planModelActivityTypes` to be consistent?

JosephVolosin replied:
Updated!

JosephVolosin committed:
Rename  =>  in derived store

JosephVolosin committed:
Add _MESSAGE to error constants

JosephVolosin committed:
Rename SequencesAndExpansionPanel => ExpansionPanel

duranb commented on src/utilities/sequence-editor/languages/handlebars/handlebars-highlighter.ts:
```diff
@@ -0,0 +1,16 @@
+import { styleTags, tags } from '@lezer/highlight';
+
+export const Highlight = styleTags({
```
Can this be exported with a name more specific to handlebars like `handlebarsHighlighter`?

JosephVolosin replied:
Renamed to `HandlebarsHighlighter`!

mattdailis replied:
I went ahead and made it lowercase 👍

duranb commented on src/utilities/permissions.ts:
```diff
@@ -1583,6 +1622,19 @@ const featurePermissions: FeaturePermissions = {
+    canCreate: user => queryPermissions.CREATE_SEQUENCE_TEMPLATE(user),
+    canDelete: (user, sequenceTemplate) => queryPermissions.DELETE_SEQUENCE_TEMPLATE(user, sequenceTemplate),
+    canRead: () => true,
+    canTemplate: (user, plan, model) => queryPermissions.EXPAND_TEMPLATES(user, plan, model),
```
Can this also be renamed to be a little more clear as to what the functionality is that is slightly more user facing? Maybe `canExpandTemplate` if "expand" is what the user sees.

JosephVolosin replied:
Renamed to `canExpand` to align w. the other expansion option

duranb commented on src/utilities/gql.ts:
```diff
@@ -2521,6 +2604,38 @@ const gql = {
 
+  SUB_MOST_RECENT_EXPANSION_FOR_SIMULATION_SEQS: `#graphql
+    subscription SubMostRecentExpansion {
+      expanded_sequences {
```
Do you mind using the existing `Queries` enums for these subscription queries?

JosephVolosin replied:
Updated, our new queries should all be using `Queries` now

mattdailis replied:
I believe this comment has been addressed, so I'll resolve it. I'm going to read this code a bit more carefully and make a new, unrelated comment.

JosephVolosin committed:
Rename Highlight => HandlebarsHighlighter

JosephVolosin committed:
Rename canTemplate => canExpand

JosephVolosin committed:
Use Queries enum for new GQL subscriptions

JosephVolosin committed:
Reorganize types

SequenceFilter is no longer dependent on ActivityLayerFilter, and common types for Layer/Sequence filters are moved to common areas

mattdailis committed:
Rename HandlebarsHighlighter to handlebarsHighlighter

mattdailis commented on src/components/expansion/ExpansionPanel.svelte:
```diff
@@ -1,351 +1,409 @@
-  let hasExpandPermission: boolean = false;
-  let seqIdInput: string = '';
-  let selectedExpansionSet: ExpansionSet | null;
+  const deletePermissionSequenceError: string = 'You do not have permission to delete an expansion sequence.';
+  const deletePermissionSequenceFilterError: string = 'You do not have permission to delete a sequence filter';
```
I think we might have a convention for where to put these messages 🤔 Running out of time now, but I can look for where they should go tomorrow.

mattdailis commented on src/components/expansion/ExpansionPanel.svelte:
```diff
@@ -1,351 +1,409 @@
-        }}
-      />
-    </PanelHeaderActions>
+    <GridMenu {gridSection} title="Sequences & Expansion" />
```
Given that we've now come full circle on the naming, I wonder how beneficial this new name is

JosephVolosin replied:
The name is not _technically false_ - creating sequences themselves is a larger part of the panel now w. filters than previously, but also there isn't really a use for sequences outside the context of expansion. I could go either way on it

mattdailis replied:
In the interest of minimizing changes that we don't have strong justification for, let's change it back to Expansion. If we feel like Expansion alone is a confusing term, let's address that separately.

I'm going over the code with an eye for things that might be considered "arbitrary" changes - things that didn't need to change in order to enable sequence templates

mattdailis commented on src/components/menus/GridMenu.svelte:
```diff
@@ -100,6 +96,10 @@
       <ExternalEventIcon />
       Selected External Event
     </MenuItem>
+    <MenuItem on:click={() => onClickMenuItem('ExpansionPanel')}>
```
Now that we've reverted the panel name, it looks like all we've done is move it to a different spot in the list. Let's move it back to where it was.

JosephVolosin replied:
Agreed, moved it back!

mattdailis commented on src/components/modals/TimeRangeModal.svelte:
```diff
@@ -0,0 +1,73 @@
+  import ModalFooter from './ModalFooter.svelte';
+  import ModalHeader from './ModalHeader.svelte';
+
+  export let height: number = 225;
+  export let width: number = 400;
```
These magic numbers jumped out at me - maybe they're fine, but I thought I should call them out and make sure this is kosher

duranb replied:
They get applied to the modal on line 42. What this does is expose these as props to the consuming component/logic in case there are certain situations where a different size is needed. Pragmatically these never change, but conventionally we do always expose these as props so I'm okay with keeping these as is.

mattdailis commented on src/components/plan/PlanMergeReview.test.ts:
```diff
@@ -13,7 +13,13 @@ import type {
 import PlanMergeReview from './PlanMergeReview.svelte';
 
-vi.mock('$env/dynamic/public', () => import.meta.env); // https://github.com/sveltejs/kit/issues/8180
+vi.mock('$env/dynamic/public', () => {
+  return {
+    env: {
+      PUBLIC_COMMAND_EXPANSION_MODE: 'typescript',
+    },
+  };
+}); // https://github.com/sveltejs/kit/issues/8180
```
I lost track of what changes happened when - are these mocks still necessary after we separated sequence templates into their own suite?

duranb replied:
Because this test is a unit test, it's using vitest as the test runner, instead of playwright. This allows the tests to affect the UI a little more dynamically over the course of the test via these mocks, whereas playwright requires the UI to be built prior to running the e2e test.

JosephVolosin committed:
Move 'Expansion' panel in plan menu back to original location

JosephVolosin committed:
TODO: Rename 'Sequences & Expansion' to 'Expansion'

mattdailis commented on package.json:
```diff
@@ -87,7 +87,8 @@
     "tippy.js": "^6.3.7",
-    "toastify-js": "^1.12.0"
+    "toastify-js": "^1.12.0",
+    "@lezer-unofficial/printer": "^1.0.1"
```
This seems to be an unused dependency. I believe David used it for debugging - we can always add it back later if we want to do some more debugging.
```suggestion
```

JosephVolosin replied:
Good catch! Removed

duranb commented on src/constants/command-expansion.ts:
```diff
@@ -0,0 +1,9 @@
+export const SEQUENCE_EXPANSION_MODE =
+  env.PUBLIC_COMMAND_EXPANSION_MODE === 'templating' ? SequencingMode.TEMPLATING : SequencingMode.TYPESCRIPT;
+export const TYPESCRIPT_EXPANSION_NOT_AVAILABLE_MESSAGE =
+  'Typescript Expansion functionality not available in "templating" command expansion mode. Please change the COMMAND_EXPANSION_MODE environment variable to "typescript" in docker-compose.yml/.env.';
```
Do you mind changing `COMMAND_EXPANSION_MODE` to `PUBLIC_COMMAND_EXPANSION_MODE` to avoid any potential confusion as to what env var needs to be changed for both these messages?

JosephVolosin replied:
Updated!

duranb commented on src/enums/filter.ts:
```diff
@@ -17,3 +8,12 @@ export enum FilterOperator {
 }
+
+export enum ActivityFilterField {
+  'Type' = 'Type',
```
Technically, we don't need the quotes around the enum names, just the values. I know there are precedents where that is the case in the existing code, so I'll leave it up to you if you want to change it; but for the most part our enums don't have the quotes.

JosephVolosin replied:
Makes sense, I went ahead and removed it from the two enums in the new `filter.ts` file

duranb commented on src/enums/sequencing.ts:
```diff
@@ -2,3 +2,8 @@ export enum SequenceTypes {
 }
+
+export enum SequencingMode {
+  TYPESCRIPT,
```
Do you mind adding explicit values to these enums to be a little more readable?

JosephVolosin replied:
Updated!

duranb commented on src/utilities/gql.ts:
```diff
@@ -2521,6 +2604,38 @@ const gql = {
+
+  SUB_MOST_RECENT_EXPANSION_FOR_SIMULATION_SIMS: `#graphql
+    subscription SubMostRecentExpansion($planId: Int!) {
+      plan_by_pk(id: $planId) {
```
There is a `Queries.PLAN` enum that can be used here instead of hardcoding `plan_by_pk`

JosephVolosin committed:
Remove unused dependency

JosephVolosin committed:
Add PUBLIC_ prefix to COMMAND_EXPANSION_MODE messages

JosephVolosin committed:
Remove extraneous quotes from enum names

JosephVolosin committed:
Set explicit values on sequencing mode enum

JosephVolosin committed:
plan_by_pk => Queries.PLAN

JosephVolosin committed:
Add TODO note for fine-grained permissions on EXPAND_TEMPLATES

duranb reviewed (approved):
looks good! great work!

JosephVolosin committed:
Linting fix

mattdailis committed:
Merge branch 'develop' into exploration/sequence-definitions
